File: scripts/g0-standalone-repro-scanner.mjs
import { constants } from 'node:fs';
import { createHash } from 'node:crypto';
import { spawn } from 'node:child_process';
import { lstat, open, readFile, realpath } from 'node:fs/promises';
import { dirname, isAbsolute, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { canonicalProviderToolJson } from './g0-provider-tool-bundle.mjs';

const PYTHON = Object.freeze({ path: '/usr/bin/python3.12', realpath: '/usr/bin/python3.12', mode: 0o755, uid: 0, sha256: '1643dacd9feaedc58f3cc581e4d22577dfe25c09b10282936186ccf0f2e61118', version: 'Python 3.12.3' });
const HELPER = Object.freeze({ path: fileURLToPath(new URL('./g0-standalone-repro-scanner.py', import.meta.url)), mode: 0o644, sha256: '94a7d1c10981ae9e877b9453f40744887094399c4fa257234c2712108bba83c0' });
const MAX_HELPER_BYTES = 64 * 1024;
const MAX_OUTPUT_BYTES = 4 * 1024 * 1024;
const DIR_FLAGS = constants.O_RDONLY | constants.O_DIRECTORY | constants.O_NOFOLLOW;
const FILE_FLAGS = constants.O_RDONLY | constants.O_NOFOLLOW;
const SHA = /^sha256:[a-f0-9]{64}$/u;
const REVISION = /^[a-f0-9]{40}$/u;
const fail = (code) => { const error = new Error(code); error.code = code; throw error; };
const identity = (st) => [st.dev, st.ino, st.mode, st.nlink, st.uid, st.gid, st.size, st.mtimeNs, st.ctimeNs].map(String).join(':');
const plain = (x) => x !== null && typeof x === 'object' && !Array.isArray(x) && Object.getPrototypeOf(x) === Object.prototype;
function exact(value, keys) { if (!plain(value) || Object.keys(value).sort().join('\0') !== [...keys].sort().join('\0')) fail('STANDALONE_REPORT_INVALID'); }
async function hashHandle(handle, size, limit) {
  if (size < 0n || size > BigInt(limit)) fail('STANDALONE_PIN_INVALID');
  const hash = createHash('sha256'); const buffer = Buffer.allocUnsafe(64 * 1024); let offset = 0;
  while (offset < Number(size)) { const { bytesRead } = await handle.read(buffer, 0, Math.min(buffer.length, Number(size) - offset), offset); if (!bytesRead) fail('STANDALONE_PIN_CHANGED'); hash.update(buffer.subarray(0, bytesRead)); offset += bytesRead; }
  return hash.digest('hex');
}
async function holdFile(pin, limit, requireRealpath = false) {
  const named = await lstat(pin.path, { bigint: true }).catch(() => fail('STANDALONE_PIN_INVALID'));
  if (!named.isFile() || named.isSymbolicLink() || named.nlink !== 1n || Number(named.mode & 0o7777n) !== pin.mode || named.uid !== BigInt(pin.uid ?? process.getuid())
      || (requireRealpath && await realpath(pin.path).catch(() => '') !== pin.realpath)) fail('STANDALONE_PIN_INVALID');
  const handle = await open(pin.path, FILE_FLAGS).catch(() => fail('STANDALONE_PIN_INVALID'));
  try { const held = await handle.stat({ bigint: true }); if (identity(named) !== identity(held) || await hashHandle(handle, held.size, limit) !== pin.sha256) fail('STANDALONE_PIN_CHANGED'); return { handle, identity: identity(held) }; }
  catch (error) { await handle.close(); throw error; }
}
async function verifyHeld(pin, held, requireRealpath = false) {
  const named = await lstat(pin.path, { bigint: true }).catch(() => fail('STANDALONE_PIN_CHANGED')); const current = await held.handle.stat({ bigint: true }).catch(() => fail('STANDALONE_PIN_CHANGED'));
  if (identity(named) !== held.identity || identity(current) !== held.identity || (requireRealpath && await realpath(pin.path).catch(() => '') !== pin.realpath)) fail('STANDALONE_PIN_CHANGED');
}
function validateOutput(value) {
  exact(value, ['contentReport', 'regularFileIdentities']); const report = value.contentReport;
  exact(report, ['artifactId', 'canonicalSourceSnapshotSha256', 'counts', 'memberHashes', 'memberInventory', 'provider', 'publicationId', 'schemaVersion', 'sourceRevision', 'tree', 'treeSha256']);
  if (report.schemaVersion !== 'wordle-royale-g0-standalone-repro-scan/v1' || !['vercel', 'railway', 'supabase'].includes(report.provider)
      || typeof report.artifactId !== 'string' || !report.artifactId.startsWith(`${report.provider}-`) || typeof report.publicationId !== 'string'
      || !SHA.test(report.canonicalSourceSnapshotSha256) || !SHA.test(report.treeSha256) || !REVISION.test(report.sourceRevision)) fail('STANDALONE_REPORT_INVALID');
  exact(report.counts, ['nodeCount', 'packageCount', 'payloadBytes']);
  if (Object.values(report.counts).some((x) => !Number.isSafeInteger(x) || x < 0)) fail('STANDALONE_REPORT_INVALID');
  const members = ['COMMIT', 'acquisition-record.json', 'bundle.tree-manifest.json', 'descriptor.json', 'install-plan.json', 'publication-index.json'];
  exact(report.memberHashes, members); if (Object.values(report.memberHashes).some((x) => !SHA.test(x))) fail('STANDALONE_REPORT_INVALID');
  if (!Array.isArray(report.tree) || report.tree.length !== report.counts.nodeCount || !Array.isArray(report.memberInventory) || report.memberInventory.length !== 6
      || !Array.isArray(value.regularFileIdentities) || value.regularFileIdentities.length !== report.tree.filter((x) => x?.type === 'file').length + 6
      || new Set(value.regularFileIdentities).size !== value.regularFileIdentities.length || value.regularFileIdentities.some((x) => !/^[0-9]+:[0-9]+$/u.test(x))) fail('STANDALONE_REPORT_INVALID');
  for (const entry of [...report.tree, ...report.memberInventory]) {
    if (!plain(entry) || typeof entry.path !== 'string' || !Number.isSafeInteger(entry.mode) || !['file', 'directory'].includes(entry.type)) fail('STANDALONE_REPORT_INVALID');
    if (entry.type === 'file' && (!Number.isSafeInteger(entry.size) || entry.size < 0 || !SHA.test(entry.sha256))) fail('STANDALONE_REPORT_INVALID');
  }
  return value;
}
function execute(python, helper, parent, publicationName) {
  return new Promise((accept, reject) => {
    const child = spawn('/proc/self/fd/4', ['/proc/self/fd/5', '6', publicationName], { env: { LANG: 'C.UTF-8', LC_ALL: 'C.UTF-8', PATH: '/usr/bin:/bin', PYTHONDONTWRITEBYTECODE: '1', PYTHONHASHSEED: '0' }, stdio: ['ignore', 'pipe', 'pipe', 'ignore', python.handle.fd, helper.handle.fd, parent.fd], windowsHide: true });
    const stdout = []; const stderr = []; let count = 0; let overflow = false;
    const consume = (target) => (chunk) => { count += chunk.length; if (count > MAX_OUTPUT_BYTES) { overflow = true; child.kill('SIGKILL'); } else target.push(Buffer.from(chunk)); };
    child.stdout.on('data', consume(stdout)); child.stderr.on('data', consume(stderr)); child.once('error', reject);
    child.once('close', (code, signal) => { if (overflow) return reject(Object.assign(new Error('STANDALONE_OUTPUT_LIMIT'), { code: 'STANDALONE_OUTPUT_LIMIT' })); if (code !== 0 || signal !== null) return reject(Object.assign(new Error(Buffer.concat(stderr).toString('utf8').trim() || 'STANDALONE_SCAN_FAILED'), { code: 'STANDALONE_SCAN_FAILED' })); accept(Buffer.concat(stdout)); });
  });
}
export async function scanProviderBundlePublicationStandalone(input) {
  exact(input, ['publicationName', 'publicationParent']);
  if (typeof input.publicationParent !== 'string' || !isAbsolute(input.publicationParent) || resolve(input.publicationParent) !== input.publicationParent || typeof input.publicationName !== 'string' || !input.publicationName || input.publicationName.includes('/')) fail('STANDALONE_INPUT_INVALID');
  const parent = await open(input.publicationParent, DIR_FLAGS).catch(() => fail('STANDALONE_PARENT_INVALID'));
  const python = await holdFile(PYTHON, 16 * 1024 * 1024, true); const helper = await holdFile(HELPER, MAX_HELPER_BYTES);
  try {
    const bytes = await execute(python, helper, parent, input.publicationName);
    if (bytes.length < 3 || bytes.length > MAX_OUTPUT_BYTES) fail('STANDALONE_OUTPUT_INVALID');
    let value; try { value = JSON.parse(bytes.toString('utf8')); } catch { fail('STANDALONE_OUTPUT_INVALID'); }
    if (!Buffer.from(`${canonicalProviderToolJson(value)}\n`).equals(bytes)) fail('STANDALONE_OUTPUT_NONCANONICAL'); validateOutput(value);
    await verifyHeld(PYTHON, python, true); await verifyHeld(HELPER, helper);
    const r = value.contentReport;
    return Object.freeze({ contentReport: r, regularFileIdentities: Object.freeze([...value.regularFileIdentities]), report: Object.freeze({ status: 'PUBLICATION_VALID', publicationValid: true, provider: r.provider, artifactId: r.artifactId, publicationId: r.publicationId, memberHashes: Object.freeze({ ...r.memberHashes }), treeSha256: r.treeSha256, canonicalSourceSnapshotSha256: r.canonicalSourceSnapshotSha256, sourceRevision: r.sourceRevision, counts: Object.freeze({ ...r.counts }) }) });
  } finally { await Promise.allSettled([parent.close(), python.handle.close(), helper.handle.close()]); }
}

export const STANDALONE_REPRO_SCANNER_TOOLCHAIN = Object.freeze({ python: Object.freeze({ version: PYTHON.version, sha256: `sha256:${PYTHON.sha256}` }), helper: Object.freeze({ sha256: `sha256:${HELPER.sha256}` }) });

File: scripts/g0-standalone-repro-scanner.py
#!/usr/bin/python3
"""Independent, descriptor-relative AN-5 publication scanner (Python stdlib only)."""
import hashlib
import json
import os
import re
import stat
import sys

MAX_JSON = 256 * 1024
MAX_MANIFEST = 2 * 1024 * 1024
MAX_FILE = 224 * 1024 * 1024
MAX_NODES = 8500
MAX_OUTPUT = 4 * 1024 * 1024
MEMBERS = ["COMMIT", "acquisition-record.json", "bundle", "bundle.tree-manifest.json", "descriptor.json", "install-plan.json", "publication-index.json"]
FILES = ["COMMIT", "acquisition-record.json", "bundle.tree-manifest.json", "descriptor.json", "install-plan.json", "publication-index.json"]
ARTIFACTS = {"vercel-58.4.4": "vercel", "railway-5.30.1": "railway", "supabase-2.110.0": "supabase"}


def fail(code):
    raise RuntimeError(code)


def canonical(value):
    return json.dumps(value, ensure_ascii=False, allow_nan=False, separators=(",", ":"), sort_keys=True).encode("utf-8") + b"\n"


def component(value):
    return isinstance(value, str) and value not in ("", ".", "..") and "/" not in value and "\0" not in value and len(value.encode()) <= 255


def safe_path(value):
    if value == ".":
        return True
    return isinstance(value, str) and not value.startswith("/") and "\\" not in value and "\0" not in value and all(component(x) for x in value.split("/")) and len(value.encode()) <= 1024


def open_dir(name, parent_fd, mode, device, uid):
    before = os.stat(name, dir_fd=parent_fd, follow_symlinks=False)
    if not stat.S_ISDIR(before.st_mode) or stat.S_ISLNK(before.st_mode) or stat.S_IMODE(before.st_mode) != mode or before.st_dev != device or before.st_uid != uid:
        fail("STANDALONE_DIRECTORY_POLICY")
    fd = os.open(name, os.O_RDONLY | os.O_DIRECTORY | os.O_NOFOLLOW, dir_fd=parent_fd)
    after = os.fstat(fd)
    if (before.st_dev, before.st_ino, before.st_mode, before.st_uid) != (after.st_dev, after.st_ino, after.st_mode, after.st_uid):
        os.close(fd); fail("STANDALONE_FILESYSTEM_CHANGED")
    return fd


def read_file(name, parent_fd, relative, mode, device, uid, limit):
    before = os.stat(name, dir_fd=parent_fd, follow_symlinks=False)
    if not stat.S_ISREG(before.st_mode) or stat.S_ISLNK(before.st_mode) or stat.S_IMODE(before.st_mode) != mode or before.st_dev != device or before.st_uid != uid or before.st_nlink != 1 or before.st_size < 0 or before.st_size > limit:
        fail("STANDALONE_FILE_POLICY")
    fd = os.open(name, os.O_RDONLY | os.O_NOFOLLOW, dir_fd=parent_fd)
    try:
        held = os.fstat(fd)
        if (before.st_dev, before.st_ino, before.st_mode, before.st_size, before.st_mtime_ns, before.st_ctime_ns) != (held.st_dev, held.st_ino, held.st_mode, held.st_size, held.st_mtime_ns, held.st_ctime_ns):
            fail("STANDALONE_FILESYSTEM_CHANGED")
        digest = hashlib.sha256(); chunks = []; remaining = held.st_size
        while remaining:
            chunk = os.read(fd, min(65536, remaining))
            if not chunk: fail("STANDALONE_FILESYSTEM_CHANGED")
            chunks.append(chunk); digest.update(chunk); remaining -= len(chunk)
        after = os.fstat(fd)
        if (held.st_dev, held.st_ino, held.st_mode, held.st_size, held.st_mtime_ns, held.st_ctime_ns) != (after.st_dev, after.st_ino, after.st_mode, after.st_size, after.st_mtime_ns, after.st_ctime_ns):
            fail("STANDALONE_FILESYSTEM_CHANGED")
        return {"path": relative, "type": "file", "mode": mode, "size": held.st_size, "sha256": "sha256:" + digest.hexdigest()}, f"{held.st_dev}:{held.st_ino}", b"".join(chunks)
    finally:
        os.close(fd)


def strict_json(raw, cap):
    if len(raw) < 3 or len(raw) > cap: fail("STANDALONE_JSON_INVALID")
    try: value = json.loads(raw.decode("utf-8"))
    except Exception: fail("STANDALONE_JSON_INVALID")
    if canonical(value) != raw: fail("STANDALONE_JSON_NONCANONICAL")
    return value


def list_names(fd):
    names = os.listdir(fd)
    if len(names) != len(set(names)) or any(not component(x) for x in names): fail("STANDALONE_PATH_INVALID")
    return sorted(names, key=lambda x: x.encode())


def scan_bundle(fd, device, uid):
    entries, detailed, identities = [], [], []
    payload = 0
    def visit(directory_fd, relative):
        nonlocal payload
        names_before = list_names(directory_fd)
        st_before = os.fstat(directory_fd)
        entries.append({"path": relative, "type": "directory", "mode": 0o555})
        detailed.append({"path": relative, "type": "directory", "mode": 0o555})
        if len(entries) > MAX_NODES: fail("STANDALONE_NODE_LIMIT")
        for name in names_before:
            rel = name if relative == "." else relative + "/" + name
            if not safe_path(rel): fail("STANDALONE_PATH_INVALID")
            st = os.stat(name, dir_fd=directory_fd, follow_symlinks=False)
            if stat.S_ISLNK(st.st_mode) or st.st_dev != device: fail("STANDALONE_NODE_POLICY")
            if stat.S_ISDIR(st.st_mode):
                child = open_dir(name, directory_fd, 0o555, device, uid)
                try: visit(child, rel)
                finally: os.close(child)
            elif stat.S_ISREG(st.st_mode):
                wanted = stat.S_IMODE(st.st_mode)
                if wanted not in (0o444, 0o555): fail("STANDALONE_FILE_POLICY")
                item, inode, _ = read_file(name, directory_fd, rel, wanted, device, uid, MAX_FILE)
                entries.append({k: item[k] for k in ("path", "type", "mode", "sha256")})
                detailed.append(item); identities.append(inode); payload += item["size"]
                if payload > 224 * 1024 * 1024: fail("STANDALONE_PAYLOAD_LIMIT")
            else: fail("STANDALONE_NODE_POLICY")
            if len(entries) > MAX_NODES: fail("STANDALONE_NODE_LIMIT")
        if names_before != list_names(directory_fd) or (st_before.st_dev, st_before.st_ino, st_before.st_mode, st_before.st_mtime_ns, st_before.st_ctime_ns) != tuple(getattr(os.fstat(directory_fd), x) for x in ("st_dev", "st_ino", "st_mode", "st_mtime_ns", "st_ctime_ns")):
            fail("STANDALONE_FILESYSTEM_CHANGED")
    visit(fd, ".")
    entries.sort(key=lambda x: x["path"].encode()); detailed.sort(key=lambda x: x["path"].encode())
    package_count = sum(1 for x in entries if x["type"] == "directory" and x["path"] != "." and ((lambda p: len(p) >= 2 and p[-2] == "node_modules" and not p[-1].startswith("@"))(x["path"].split("/")) or (lambda p: len(p) >= 3 and p[-3] == "node_modules" and p[-2].startswith("@") and not p[-1].startswith("@"))(x["path"].split("/"))))
    return entries, detailed, identities, package_count, payload


def main():
    if len(sys.argv) != 3 or not sys.argv[1].isdigit() or not component(sys.argv[2]): fail("STANDALONE_INPUT_INVALID")
    parent_fd = int(sys.argv[1]); publication_name = sys.argv[2]
    parent = os.fstat(parent_fd); uid = os.getuid(); device = parent.st_dev
    container = open_dir(publication_name, parent_fd, 0o700, device, uid)
    try:
        if list_names(container) != MEMBERS: fail("STANDALONE_MEMBERS_INVALID")
        member_hashes, member_inventory, identities, documents = {}, [], [], {}
        for name in FILES:
            item, inode, raw = read_file(name, container, name, 0o400, device, uid, MAX_MANIFEST if name == "bundle.tree-manifest.json" else MAX_JSON)
            member_hashes[name] = item["sha256"]; member_inventory.append(item); identities.append(inode); documents[name] = strict_json(raw, MAX_MANIFEST if name == "bundle.tree-manifest.json" else MAX_JSON)
        bundle_fd = open_dir("bundle", container, 0o555, device, uid)
        try: entries, tree, tree_ids, package_count, payload = scan_bundle(bundle_fd, device, uid)
        finally: os.close(bundle_fd)
        identities.extend(tree_ids)
        if len(identities) != len(set(identities)): fail("STANDALONE_HARDLINK_FORBIDDEN")
        manifest = documents["bundle.tree-manifest.json"]
        if not isinstance(manifest, dict) or manifest.get("entries") != entries: fail("STANDALONE_MANIFEST_MISMATCH")
        index = documents["publication-index.json"]; acquisition = documents["acquisition-record.json"]; commit = documents["COMMIT"]
        artifact = index.get("artifactId") if isinstance(index, dict) else None
        provider = ARTIFACTS.get(artifact)
        index_hash = member_hashes["publication-index.json"]
        derived = f"{artifact}-{index_hash[7:39]}"
        if provider is None or publication_name != derived or commit.get("publicationIndexSha256") != index_hash or acquisition.get("canonicalSourceSnapshotSha256") != index.get("canonicalSourceSnapshotSha256"):
            fail("STANDALONE_BINDING_INVALID")
        report = {"artifactId": artifact, "canonicalSourceSnapshotSha256": index["canonicalSourceSnapshotSha256"], "counts": {"nodeCount": len(entries), "packageCount": package_count, "payloadBytes": payload}, "memberHashes": member_hashes, "memberInventory": member_inventory, "provider": provider, "publicationId": derived, "schemaVersion": "wordle-royale-g0-standalone-repro-scan/v1", "sourceRevision": index.get("sourceRevision"), "tree": tree, "treeSha256": member_hashes["bundle.tree-manifest.json"]}
        result = {"contentReport": report, "regularFileIdentities": sorted(identities, key=lambda x: tuple(map(int, x.split(":"))))}
        output = canonical(result)
        if len(output) > MAX_OUTPUT: fail("STANDALONE_OUTPUT_LIMIT")
        os.write(1, output)
    finally: os.close(container)


if __name__ == "__main__":
    try: main()
    except Exception as error:
        code = str(error)
        os.write(2, (code if re.fullmatch(r"[A-Z0-9_]+", code) else "STANDALONE_SCAN_FAILED").encode() + b"\n")
        raise SystemExit(1)
